feat: adding ALZ Policy Versioning gudiance (#66)
Azure/Azure-Landing-Zones: docs/content/policy/policyVersioning.md

---
title: ALZ and Policy Versioning
weight: 10
---

{{< hint type=important >}}
Even though still in public preview, it is highly recommended to implement Azure Policy versioning and update policies/initiatives and assignments by pinning to major versions as soon as possible to avoid unintended consequences due to breaking changes being released.
{{< /hint >}}

In mid-2024 the Azure Policy product group released a new versioning scheme for Azure Policy definitions. This new versioning scheme is designed to provide a more consistent and predictable way to manage policy definitions and initiatives, making it easier for organizations to adopt and implement Azure Policy.

In the context of Azure Policy, versioning refers to the practice of managing and maintaining different versions of policy definitions and initiatives. This is important for several reasons:

- **Backward Compatibility**: As Azure Policy evolves, new features and capabilities are introduced. Versioning ensures that existing policies continue to work as expected, even as new versions are released.

> - **Change Management**: Versioning allows organizations to track changes made to policies over time. This is crucial for auditing and compliance purposes.

Author: @paulgrimley you can track changes in AzAdvertizer for example, and see the changes to policy over time.

- **Testing and Validation**: When a new version of a policy is released, organizations can test it in a controlled environment before deploying it to production. This helps identify any potential issues or conflicts with existing policies.

> - **Documentation**: Versioning provides a clear history of changes made to policies, making it easier for teams to understand the evolution of their policy landscape.

Reviewer: Perhaps refer to AzAdvertizer for this as i am not sure its easy natively to determine release history and this is the value add ALZ can share

- **Rollback Capability**: If a new version of a policy causes issues, versioning allows organizations to revert to a previous version quickly.
- **Feature Deprecation**: As Azure Policy matures, certain features may be deprecated or replaced with better alternatives. Versioning allows organizations to manage the transition to new features while maintaining support for older versions.

> - **Community Contributions**: In open-source or community-driven environments, versioning allows contributors to submit changes and improvements without disrupting existing policies.

Reviewer: If this is related to custom policies then we need to remove this from here as i dont think this is valid

Author: @paulgrimley our custom initiatives reference built-in policies, and now need to specify the version of the policy to use. But will remove as I can see this may be confusing.

- **Integration with CI/CD**: Versioning policies can be integrated into continuous integration and continuous deployment (CI/CD) pipelines, allowing for automated testing and deployment of policy changes.

{{< hint type=note >}}

> As of March 2025, Azure Policy version only supports built-in policies and initiatives. Custom policies and initiatives are not supported.

Reviewer: ```suggestion
As of March 2025, Azure Policy version only supports built-in policies and initiatives. Custom policies and initiatives are not supported but are planned.
```

{{< /hint >}}

## How it works

> The new versioning scheme for Azure Policy is based on semantic versioning, which consists of three components: major, minor, and patch versions. Each component serves a specific purpose:

Reviewer: ```suggestion
The new versioning scheme for Azure Policy is based on [semantic versioning](https://semver.org/), which consists of three components: major, minor, and patch versions. Each component serves a specific purpose:
```

{{< hint type=note >}}
The `version` attribute under `metadata` is not used for Azure Policy versioning. It is used for other purposes, such as tracking changes to the policy definition over time, and should not be confused with the new versioning scheme. It serves no functional purpose in the Azure Policy engine.
{{< /hint >}}

## Policy Versioning in ALZ

Azure Policy versioning introduced several benefits but also required ALZ to adapt to the new framework. To align with this change, the ALZ team implemented versioning for built-in policies and initiatives by pinning them to their current major version within custom initiatives and assignments. This ensures ALZ consistently uses the latest minor and patch versions while maintaining a stable major version (e.g., 1.x). As a result, users benefit from the most up-to-date and validated version of the pinned major release.

This update was published as part of the **Q2 FY25 Policy Refresh (January 2025)**, during which all custom initiatives and assignments referencing built-in policies and initiatives were updated to align with the latest available major version.

> By pinning to the current major version, ALZ does not automatically upgrade to newly released major versions. This approach allows the ALZ team time to review and assess any breaking changes before initiating an upgrade.

Reviewer: What should customers do if they want to move to a new version ahead of an ALZ release (and include risks)?

Author: @paulgrimley I don't know how to answer this. The whole point of this page is to explain our approach to helping customers adopt policy through ALZ. If they want to do their own thing, then go look at Learn docs and accept all the risks and responsibilities accordingly. We're not here to teach people how to use policy.

Moving forward, the ALZ team will monitor new major versions of built-in policies and initiatives used by ALZ. Updates will be incorporated as part of the regular policy refresh cycle once the necessary changes have been reviewed, tested, and validated.

{{< hint type=note >}}

> Due to upstream dependencies on product SDKs, Azure Policy versioning is not yet supported in the Terraform and Bicep accelerators.

Reviewer: ```suggestion
Due to upstream dependencies on product SDKs, Azure Policy versioning is not yet supported in Terraform Azure Verified Modules for Platform Landing Zone (ALZ).
```
Versioning works in the current Bicep version (if SDK not addressed before next Bicep release then this will become an issue but hoping this wont be the case)

{{< /hint >}}

## Updating

Depending on the accelerator you've used to deploy your Azure Landing Zones, you may need to update your policies/initiatives and assignments.

> - **Portal**: If you have deployed your Azure Landing Zones using the portal, you will need to update your policies/initiatives and assignments manually. This is because the portal does not support automatic updates for policies/initiatives and assignments.

Reviewer: Do we have guidance for this? if not perhaps we need to create something, thoughts?

Author: @paulgrimley yes, see the link below.

> - **Bicep**: If you have deployed your Azure Landing Zones using Bicep, you will need to update your policies/initiatives and assignments manually. This is because Bicep does not support automatic updates for policies/initiatives and assignments.

Reviewer: Are you sure? as part of updates to Bicep dont we add changes to policies, check with Zach

Author: New policies will be ingested from the policy library, and will subsequently be deployed, but it still requires manual steps to update the assignment

> - **Terraform**: If you have deployed your Azure Landing Zones using Terraform, you will need to update to the latest version of the accelerator and apply the changes using your normal Terraform workflow.

Reviewer: I dont think its the accelerator we need to update its the AVM modules - please check with @jaredfholgate

- **Documentation**: Versioning provides a clear history of changes made to policies, making it easier for teams to understand the evolution of their policy landscape.
- **Rollback Capability**: If a new version of a policy causes issues, versioning allows organizations to revert to a previous version quickly.
- **Feature Deprecation**: As Azure Policy matures, certain features may be deprecated or replaced with better alternatives. Versioning allows organizations to manage the transition to new features while maintaining support for older versions.
- **Integration with CI/CD**: Versioning policies can be integrated into continuous integration and continuous deployment (CI/CD) pipelines, allowing for automated testing and deployment of policy changes.

{{< hint type=note >}}
As of March 2025, Azure Policy version only supports built-in policies and initiatives. Custom policies and initiatives are not supported but are planned.
{{< /hint >}}

## How it works

> The new versioning scheme for Azure Policy is based on semantic versioning, which consists of three components: major, minor, and patch versions. Each component serves a specific purpose:

Reviewer 2: Looks like line 28 and 29 are duplicated.

Author: GitHub fail

- **Patch Version**: Indicates bug fixes or minor changes that do not affect the policy's functionality. A change in the patch version number signifies that the new version includes fixes but does not introduce new features or breaking changes.

The most important thing to note is that the major version number is incremented when breaking changes are introduced. This means that if you are using a policy definition with a specific major version, you should be aware that upgrading to a new major version may require changes to your implementation.

All Azure built-in policies and initiatives are now versioned using the new scheme. The version number is included in the policy definition's metadata, making it easy to identify the version being used and is typically formatted as `X.Y.Z` where `X` is the major version, `Y` is the minor version, and `Z` is the patch version, e.g. `1.*.*`. The `*` are used for minor and patch version as wildcards in initiatives and assignments, as changes to these versions are backward compatible and will be used automatically.

> {{< figure src="../img/policyWithVersion.png" title="Policy with version example">}}

Reviewer 2: Nitpick. `title="Policy with version example"` looks like a title for a new paragraph. 

![image](https://github.com/user-attachments/assets/6600f0f0-c151-4ae5-aa21-ad7b2fd117f6)
